frame/staking/src/lib.rs

```diff
@@ -1366,66 +1543,60 @@ impl<T: Trait> Module<T> {
 		Self::new_era(session_index)
 	}
 
-	/// Initialise the first session (and consequently the first era)
+	/// Initialise the session 0 (and consequently the era 0)
 	fn initial_session() -> Option<Vec<T::AccountId>> {
 		// note: `CurrentEraStart` is set in `on_finalize` of the first block because now is not
 		// available yet.
 		CurrentEraStartSessionIndex::put(0);
 		BondedEras::mutate(|bonded| bonded.push((0, 0)));
-		Self::select_validators().1
+		Self::select_validators()
 	}
 
-	/// The era has changed - enact new staking set.
-	///
-	/// NOTE: This always happens immediately before a session change to ensure that new validators
-	/// get a chance to set their session keys.
-	fn new_era(start_session_index: SessionIndex) -> Option<Vec<T::AccountId>> {
-		// Payout
-		let points = CurrentEraPointsEarned::take();
-		let now = T::Time::now();
-		let previous_era_start = <CurrentEraStart<T>>::mutate(|v| {
-			sp_std::mem::replace(v, now)
-		});
-		let era_duration = now - previous_era_start;
-		if !era_duration.is_zero() {
-			let validators = Self::current_elected();
-
-			let validator_len: BalanceOf<T> = (validators.len() as u32).into();
-			let total_rewarded_stake = Self::slot_stake() * validator_len;
-
-			let (total_payout, max_payout) = inflation::compute_total_payout(
-				&T::RewardCurve::get(),
-				total_rewarded_stake.clone(),
-				T::Currency::total_issuance(),
-				// Duration of era; more than u64::MAX is rewarded as u64::MAX.
-				era_duration.saturated_into::<u64>(),
-			);
-
-			let mut total_imbalance = <PositiveImbalanceOf<T>>::zero();
+	fn end_session(session_index: SessionIndex) {
+		if ErasStartSessionIndex::get(Self::active_era() + 1) == session_index + 1 {
```

Comment:
one tricky aspect of this condition is that if somehow one session is missed by session pallet then we are stuck in the current era forever.

Reply:
this could be fixed by some emergency situation like this
```diff
diff --git a/frame/staking/src/lib.rs b/frame/staking/src/lib.rs
index 0c8b2320d..4bd052da3 100644
--- a/frame/staking/src/lib.rs
+++ b/frame/staking/src/lib.rs
@@ -1553,7 +1553,16 @@ impl<T: Trait> Module<T> {
        }
 
        fn end_session(session_index: SessionIndex) {
-               if ErasStartSessionIndex::get(Self::active_era() + 1) == session_index + 1 {
+               let active_era = Self::active_era();
+               let current_era = Self::current_era();
+               let next_era_start = ErasStartSessionIndex::get(active_era + 1);
+
+               // In case the next_era_start has been missed and there is some era planned then we end the
+               // current era.
+               // This must never happen
+               let emergency = next_era_start < session_index + 1 && active_era < current_era;
+
+               if next_era_start == session_index + 1 || emergency {
                        Self::end_era(session_index);
                }
        }
```